metaflow/plugins/cards/card_modules/components.py

```diff
@@ -366,64 +468,84 @@ def from_matplotlib(cls, plot, label: Optional[str] = None):
         """
-        import io
-
         try:
-            try:
-                import matplotlib.pyplot as pyplt
-            except ImportError:
-                return ErrorComponent(
-                    cls.render_fail_headline("Matplotlib cannot be imported"),
-                    "%s" % traceback.format_exc(),
-                )
-            # First check if it is a valid Matplotlib figure.
-            figure = None
-            if _full_classname(plot) == "matplotlib.figure.Figure":
-                figure = plot
-
-            # If it is not valid figure then check if it is matplotlib.axes.Axes or a matplotlib.axes._subplots.AxesSubplot
-            # These contain the `get_figure` function to get the main figure object.
-            if figure is None:
-                if getattr(plot, "get_figure", None) is None:
-                    return ErrorComponent(
-                        cls.render_fail_headline(
-                            "Invalid Type. Object %s is not from `matplotlib`"
-                            % type(plot)
-                        ),
-                        "",
-                    )
-                else:
-                    figure = plot.get_figure()
-
-            task_to_dict = TaskToDict()
-            img_bytes_arr = io.BytesIO()
-            figure.savefig(img_bytes_arr, format="PNG")
-            parsed_image = task_to_dict.parse_image(img_bytes_arr.getvalue())
-            pyplt.close(figure)
+            parsed_image, error_comp = cls._parse_matplotlib(plot)
             if parsed_image is not None:
-                return cls(src=parsed_image, label=label)
-            return ErrorComponent(
-                cls.render_fail_headline("Matplotlib plot's image is not parsable"), ""
-            )
+                img = cls(src=parsed_image, label=label)
+            else:
+                img = cls(src=None, label=label)
+                img._error_comp = error_comp
+            return img
         except:
             import traceback
 
-            return ErrorComponent(
+            img = cls(src=None, label=label)
+            img._error_comp = ErrorComponent(
                 cls.render_fail_headline("Matplotlib plot's image is not parsable"),
                 "%s" % traceback.format_exc(),
             )
+            return img
 
+    @with_default_component_id
     @render_safely
     def render(self):
         if self._error_comp is not None:
             return self._error_comp.render()
 
         if self._src is not None:
-            return ImageComponent(src=self._src, label=self._label).render()
+            img_comp = ImageComponent(src=self._src, label=self._label)
+            img_comp.component_id = self.component_id
+            return img_comp.render()
         return ErrorComponent(
             self.render_fail_headline("`Image` Component `src` argument is `None`"), ""
         ).render()
 
+    def update(self, pilimage=None, plot=None, bytes=None, string=None, label=None):
```

Comment:
Image update signature; cc @tuulos

Reply:
do we need all the different kwargs? Couldn't we just have `update(image_object, label)` and then we detect internally the type of the object, update accordingly?

You have the code for checking the class for `matplotlib` etc. elsewhere in the class

Reply:
Will Fix!